src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/RazorSemanticTokensInfoService.cs

```diff
@@ -95,22 +96,40 @@ internal class RazorSemanticTokensInfoService : IRazorSemanticTokensInfoService
         return tokens;
     }
 
-    private static List<SemanticRange> CombineSemanticRanges(List<SemanticRange> ranges1, List<SemanticRange> ranges2)
+    private static ImmutableArray<SemanticRange> CombineSemanticRanges(ImmutableArray<SemanticRange> razorRanges, ImmutableArray<SemanticRange> csharpRanges)
     {
-        var newList = new List<SemanticRange>(ranges1.Count + ranges2.Count);
-        newList.AddRange(ranges1);
-        newList.AddRange(ranges2);
+        Debug.Assert(razorRanges.SequenceEqual(razorRanges.OrderBy(g => g)));
+
+        // If there are no C# in what we're trying to classify we can skip this, since we know the razor ranges will be sorted.
+        if (csharpRanges.Length == 0)
+        {
+            return razorRanges;
+        }
+
+        // If there are no Razor ranges then we can't fully fast path, as the C# ranges are not necessarily sorted, but we can
+        // avoid creating a new builder and copying all of the elements. Or maybe that's what the Sort method does internally
+        // anyway 🤷‍
+        if (razorRanges.Length == 0)
+        {
+            return csharpRanges.Sort();
+        }
+
+        using var _ = ArrayBuilderPool<SemanticRange>.GetPooledObject(out var newList);
+        newList.SetCapacityIfLarger(razorRanges.Length + csharpRanges.Length);
+
+        newList.AddRange(razorRanges);
+        newList.AddRange(csharpRanges);
 
         // Because SemanticToken data is generated relative to the previous token it must be in order.
         // We have a guarantee of order within any given language server, but the interweaving of them can be quite complex.
```

Comment:
>We have a guarantee of order within any given language server [](http://example.com/codeflow?start=11&length=61)

Is this comment not accurate (and thus the need for you to sort in the just csharp case above?)

Reply:
Strictly speaking its not wrong, but certainly misleading. Will update.

Within any given language server the order is guaranteed, due to the nature of the offset based data, but that just means its in order when we get it back from the C# server. Once we've done the re-mapping, a `using` directive which would be at the top of the generated C# file, can produce a classified span in the middle of a .razor document (as the `@using` can appear anywhere).